pkg/endpoint/bpf.go

```diff
@@ -1311,21 +1330,14 @@ func (e *Endpoint) applyPolicyMapChanges() error {
 // difference between the realized and desired policy state without
 // dumping the bpf policy map.
 func (e *Endpoint) syncPolicyMap() error {
-	// Apply pending policy map changes first so that desired map is up-to-date before
-	// we diff the maps below.
-	err := e.applyPolicyMapChanges()
```

Comment:
So, we don't need this anymore because we will apply the pending incremental updates in a subsequent incremental update, right? Thus fixing the leak where incremental updates that come in during regeneration are lost in Envoy?

Reply:
This call is moved higher up in the call stack, so that we have consistent view on the mapstate for both bpf map sync and Envoy network policy update, which by design needs to happen before the bpf map sync.
```
--- a/pkg/endpoint/bpf.go
+++ b/pkg/endpoint/bpf.go
@@ -763,6 +763,13 @@ func (e *Endpoint) runPreCompilationSteps(regenContext *regenerationContext, rul
 	// Mark the desired policy as ready when done before the lock is released
 	defer e.desiredPolicy.Ready()
 
+	// Apply pending policy map changes so that desired map is up-to-date before
+	// we update network policies and sync the maps below.
+	err = e.applyPolicyMapChanges()
+	if err != nil {
+		return err
+	}
+
 	// We cannot obtain the rules while e.mutex is held, because obtaining
 	// fresh DNSRules requires the IPCache lock (which must not be taken while
 	// holding e.mutex to avoid deadlocks). Therefore, rules are obtained
```